infoschema/cache.go

```diff
@@ -129,9 +162,31 @@ func (h *InfoCache) Insert(is InfoSchema, snapshotTS uint64) bool {
 		return h.cache[i].SchemaMetaVersion() <= version
 	})
 
-	if h.maxUpdatedSnapshotTS < snapshotTS {
-		h.maxUpdatedSnapshotTS = snapshotTS
+	// maintain ts to version mapping
+	ts, ok := h.versionToTs[version]
```

Comment:
You could use original logic, `if i < len(h.cache) && h.cache[i].SchemaMetaVersion() == version {` at L192-209.

It is a little bit chaotic with two bidirectional maps. You never clear `versionToTS` and `tsToVersion` map, in fact, it will be a leak.

I recommend another implementation:

```
cache []struct{
 infoschema Infoschema
 timestamp int64
}
sortedTS []struct {
  version int64
  timestamp int64
}
```

Every time changing `cache`, you add or remove the corresponding entry in `sortedTS`. Every time caching the same schema, find and update timestamp entry.

Reply:
oh, nice, let me change it